docs/how-to-guides/upgrade-etcd.md

```diff
@@ -0,0 +1,110 @@
+
+```bash
+ssh core@<IP Address>
+```
+
+### Step 2: Create necessary directories with correct permissions
+
+Latest etcd (`>= v3.4.10`) necessitates the data directory permissions to be `0700`, accordingly change the permissions. Verify the permissions are changed to `rwx------`.
+
+> **NOTE**: This step is needed only for the Lokomotive deployment done using `lokoctl` version `< 0.4.0`.
+
+```bash
+sudo chmod 0700 /var/lib/etcd/
+sudo ls -ld /var/lib/etcd/
+```
+
+If the node reboots, we need the right settings in place so that `systemd-tmpfile` service does not alter the permissions of the data directory. To make the changes made above persistent run the following command:
+
+```bash
+echo "d    /var/lib/etcd 0700 etcd etcd - -" | sudo tee /etc/tmpfiles.d/etcd-wrapper.conf
+```
+
+### Step 3: Upgrade etcd
+
+Run the following commands:
+
+> **NOTE**: Before proceeding to other commands, set the `etcd_version` variable to the latest etcd version.
+
+```bash
+export etcd_version=<latest etcd version e.g. v3.4.10>
+
+sudo sed -i "s,ETCD_IMAGE_TAG=.*,ETCD_IMAGE_TAG=${etcd_version}," \
+        /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf
+sudo systemctl daemon-reload
+sudo systemctl restart etcd-member
+```
+
+### Step 4: Verify upgrade
+
+Verify that the etcd service is in `active (running)` state:
+
+```bash
+sudo systemctl status --no-pager etcd-member
+```
+
+Run the following command to see logs of the process since the last restart:
+
+```bash
+sudo journalctl _SYSTEMD_INVOCATION_ID=$(sudo systemctl \
+              show -p InvocationID --value etcd-member.service)
+```
+
+> **NOTE**: Do not proceed with the upgrade of the rest of the cluster if you encounter any errors.
+
+Once you see the following log line, you can discern that the etcd daemon has come up without errors:
+
+```log
+etcdserver: starting server... [version: 3.4.10, cluster version: to_be_decided]
+```
+
+Once you see the following log line, you can discern that the etcd has rejoined the cluster without issues:
+
+```log
+embed: serving client requests on 10.88.81.1:2379
+```
+
+### Step 5: Verify using `etcdctl`
+
+We can use `etcdctl` client to verify the state of etcd cluster.
+
+```bash
+# Find the endpoint of this node's etcd:
+export endpoint=$(grep ETCD_ADVERTISE_CLIENT_URLS /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf | cut -d"=" -f3 | tr -d '"')
+
+export flags="--cacert=/etc/ssl/etcd/etcd-client-ca.crt \
+              --cert=/etc/ssl/etcd/etcd-client.crt \
+              --key=/etc/ssl/etcd/etcd-client.key \
+              --endpoints=${endpoint}"
+
+# Verify:
+sudo ETCDCTL_API=3 etcdctl member list $flags
+sudo ETCDCTL_API=3 etcdctl endpoint health $flags
+```
+
+The last command should report that node is healthy. If it indicates otherwise then try commands from [Step 4](#step-4-verify-upgrade) to see what's wrong. If the node is healthy, it is safe to move forward with the next controller node.
```

Comment:
Just curious: shouldn't the node be healthy and the first command indicate that there are 3 members? I mean, in normal scenarios if there are 2 members and the node is healthy (not sure this can happen), we wouldn't want to continue, right?

Reply:
> if there are 2 members and the node is healthy (not sure this can happen),

Do you mean unhealthy here?

Reply:
No, healthy. I mean, let's say the last command `etcdctl endpoint health` shows healthy, but the previous one about members shows 2. We don't want to continue, right? Not sure if such a thing can happen, though, as if you are healthy you might have joined the cluster and already synced etcd data?

Reply:
@rata now we have a reliable way of finding all the endpoints \o/. See the current changes.

```bash
core@mycluster-controller-2 ~ $ # Find the endpoint of this node's etcd:
core@mycluster-controller-2 ~ $ export endpoint=$(grep ETCD_ADVERTISE_CLIENT_URLS \
>         /etc/systemd/system/etcd-member.service.d/40-etcd-cluster.conf | cut -d"=" -f3 | tr -d '"')


core@mycluster-controller-2 ~ $ export flags="--cacert=/etc/ssl/etcd/etcd-client-ca.crt \
>               --cert=/etc/ssl/etcd/etcd-client.crt \
>               --key=/etc/ssl/etcd/etcd-client.key"


core@mycluster-controller-2 ~ $ endpoints=$(sudo ETCDCTL_API=3 etcdctl member list $flags --endpoints=${endpoint} \
>             --write-out=json | jq -r '.members[].clientURLs[]')

core@mycluster-controller-2 ~ $ endpoints=$(sed 's| |,|g' <<< ${endpoints})

core@mycluster-controller-2 ~ $ # Verify:
core@mycluster-controller-2 ~ $ sudo ETCDCTL_API=3 etcdctl member list $flags --endpoints=${endpoint}
2b89a567f6c2e2e8, started, etcd0, https://mycluster-etcd0.dev.mycomp.com:2380, https://mycluster-etcd0.dev.mycomp.com:2379
99588ac1031498cf, started, etcd2, https://mycluster-etcd2.dev.mycomp.com:2380, https://mycluster-etcd2.dev.mycomp.com:2379
eec240697239d5b8, started, etcd1, https://mycluster-etcd1.dev.mycomp.com:2380, https://mycluster-etcd1.dev.mycomp.com:2379


core@mycluster-controller-2 ~ $ sudo ETCDCTL_API=3 etcdctl endpoint health $flags --endpoints=${endpoints}
https://mycluster-etcd2.dev.mycomp.com:2379 is healthy: successfully committed proposal: took = 29.811666ms
https://mycluster-etcd0.dev.mycomp.com:2379 is healthy: successfully committed proposal: took = 40.142713ms
https://mycluster-etcd1.dev.mycomp.com:2379 is healthy: successfully committed proposal: took = 40.819513ms
```

Reply:
Cool!
My point still is:

> My point is: before continue we want to make sure the cluster has 3 members again and etcd data has been fully synced. Not sure if healthy implies all of that or not :) (maybe it is healthy because joined and is syncing, but haven't completed? Don't really know :))

So, I'd update the text to reflect that (if this is enough to see that etcd has 3 members and all data is synced) :)

Reply:
Data is not lost when you upgrade etcd, the data is already there. 

We can add something like this at the end and say that verify that the db size is same for all the nodes.

```console
$ sudo ETCDCTL_API=3 etcdctl endpoint status -w table $flags --endpoints=${endpoints}
+-------------------------------------+------------------+---------+---------+-----------+-----------+------------+
|               ENDPOINT              |        ID        | VERSION | DB SIZE | IS LEADER | RAFT TERM | RAFT INDEX |
+-------------------------------------+------------------+---------+---------+-----------+-----------+------------+
| https://mycluster-etcd0.mycomp:2379 | 2b89a567f6c2e2e8 |   3.4.7 |  7.6 MB |      true |         2 |      46614 |
| https://mycluster-etcd2.mycomp:2379 | 99588ac1031498cf |   3.4.7 |  7.6 MB |     false |         2 |      46614 |
| https://mycluster-etcd1.mycomp:2379 | eec240697239d5b8 |   3.4.7 |  7.7 MB |     false |         2 |      46614 |
+-------------------------------------+------------------+---------+---------+-----------+-----------+------------+
```

Reply:
Nah, seems too complicated. Also, same size doesn't mean it is in sync (a small downtime, like restaritng etcd on upgrade, won't be noticeable on mega bytes but can still missed a small change).

I guess this is fine for now. Thanks again! :)